### PR TITLE
YD-346 Fix get message concurrency issue

### PR DESCRIPTION
**Pull request by aristotelos**
- Added try/catch, using subtransaction
- Refactored some existing methods a little too

### CODE REVIEW

**Comment by aristotelos** [comment]
Warning: I was not yet able to run the integration tests because of Hazelcast issues.

**Comment by Bert-R** [comment]
Note that I synced it with the master. You will need to do a pull.

**Comment by Bert-R** [comment]
Pushed one more update: an informational log message, so we can see when (how often) this occurs.

There are two failing tests:
```
Overwrite Richard while being a buddy of Bob

Condition not satisfied:

getMessagesResponse.responseData._embedded?."yona:messages"?.size() == 1
|                   |            |          |                |      |
|                   |            null       null             null   false
|                   [_links:[self:[href:http://localhost:8082/users/81c9820d-d4e5-4224-8f60-fa2981a9251a/messages/]], page:[number:0, size:20, totalElements:0, totalPages:0]]
groovyx.net.http.HttpResponseDecorator@53d864e6

	at nu.yona.server.OverwriteUserTest.Overwrite Richard while being a buddy of Bob(OverwriteUserTest.groovy:76)

Overwrite Richard while being a buddy of Bob and verify that Bob can move on with everything

Condition not satisfied:

getMessagesBobResponse.responseData._embedded."yona:messages".size() == 2
|                      |            |         |               |      |
|                      |            |         |               1      false
|                      |            |         [[@type:GoalConflictMessage, _links:[edit:[href:http://localhost:8082/users/1f9561f5-4c91-4ce2-8a53-a3f71c83d02b/messages/6e0abcb5-1215-48f5-b110-547200acf607], self:[href:http://localhost:8082/users/1f9561f5-4c91-4ce2-8a53-a3f71c83d02b/messages/6e0abcb5-1215-48f5-b110-547200acf607], yona:activityCategory:[href:http://localhost:8082/activityCategories/192d69f4-8d3e-499b-983c-36ca97340ba9], yona:dayDetails:[href:http://localhost:8082/users/1f9561f5-4c91-4ce2-8a53-a3f71c83d02b/activity/days/2016-10-29/details/f9654017-4bd4-4452-9f95-92da5414e61f], yona:markRead:[href:http://localhost:8082/users/1f9561f5-4c91-4ce2-8a53-a3f71c83d02b/messages/6e0abcb5-1215-48f5-b110-547200acf607/markRead]], activityEndTime:2016-10-29T10:52:05.797+0200, activityStartTime:2016-10-29T10:51:05.797+0200, creationTime:2016-10-29T10:51:05.802+0200, isRead:false, nickname:BD (me), status:ANNOUNCED, url:http://www.poker.com]]
|                      |            [yona:messages:[[@type:GoalConflictMessage, _links:[edit:[href:http://localhost:8082/users/1f9561f5-4c91-4ce2-8a53-a3f71c83d02b/messages/6e0abcb5-1215-48f5-b110-547200acf607], self:[href:http://localhost:8082/users/1f9561f5-4c91-4ce2-8a53-a3f71c83d02b/messages/6e0abcb5-1215-48f5-b110-547200acf607], yona:activityCategory:[href:http://localhost:8082/activityCategories/192d69f4-8d3e-499b-983c-36ca97340ba9], yona:dayDetails:[href:http://localhost:8082/users/1f9561f5-4c91-4ce2-8a53-a3f71c83d02b/activity/days/2016-10-29/details/f9654017-4bd4-4452-9f95-92da5414e61f], yona:markRead:[href:http://localhost:8082/users/1f9561f5-4c91-4ce2-8a53-a3f71c83d02b/messages/6e0abcb5-1215-48f5-b110-547200acf607/markRead]], activityEndTime:2016-10-29T10:52:05.797+0200, activityStartTime:2016-10-29T10:51:05.797+0200, creationTime:2016-10-29T10:51:05.802+0200, isRead:false, nickname:BD (me), status:ANNOUNCED, url:http://www.poker.com]]]
|                      [_embedded:[yona:messages:[[@type:GoalConflictMessage, _links:[edit:[href:http://localhost:8082/users/1f9561f5-4c91-4ce2-8a53-a3f71c83d02b/messages/6e0abcb5-1215-48f5-b110-547200acf607], self:[href:http://localhost:8082/users/1f9561f5-4c91-4ce2-8a53-a3f71c83d02b/messages/6e0abcb5-1215-48f5-b110-547200acf607], yona:activityCategory:[href:http://localhost:8082/activityCategories/192d69f4-8d3e-499b-983c-36ca97340ba9], yona:dayDetails:[href:http://localhost:8082/users/1f9561f5-4c91-4ce2-8a53-a3f71c83d02b/activity/days/2016-10-29/details/f9654017-4bd4-4452-9f95-92da5414e61f], yona:markRead:[href:http://localhost:8082/users/1f9561f5-4c91-4ce2-8a53-a3f71c83d02b/messages/6e0abcb5-1215-48f5-b110-547200acf607/markRead]], activityEndTime:2016-10-29T10:52:05.797+0200, activityStartTime:2016-10-29T10:51:05.797+0200, creationTime:2016-10-29T10:51:05.802+0200, isRead:false, nickname:BD (me), status:ANNOUNCED, url:http://www.poker.com]]], _links:[curies:[[href:http://dev.yona.nu/rels/{rel}, name:yona, templated:true]], self:[href:http://localhost:8082/users/1f9561f5-4c91-4ce2-8a53-a3f71c83d02b/messages/]], page:[number:0, size:20, totalElements:1, totalPages:1]]
groovyx.net.http.HttpResponseDecorator@5d9cdf0f

	at nu.yona.server.OverwriteUserTest.Overwrite Richard while being a buddy of Bob and verify that Bob can move on with everything(OverwriteUserTest.groovy:213)
```

**Comment by aristotelos** [comment]
Updated code, fixing the tests.

File: core/src/main/java/nu/yona/server/messaging/service/MessageService.java [hunk 1]
```diff
@@ -54,12 +60,55 @@
...
+			tryTransferDirectMessagesToAnonymousDestination(user);
+		}
+		catch (DataIntegrityViolationException e)
+		{
+			if (e.getCause() instanceof ConstraintViolationException)
+			{
+				// Ignore and proceed. Another concurrent thread has transferred the messages.
+				// We avoid a lock here because that limits scaling this service horizontally.
+				logger.info(
+						"The direct log messages of user with mobile number '" + user.getMobileNumber() + "' and ID '"
```
Comment (Bert-R) [inline core/src/main/java/nu/yona/server/messaging/service/MessageService.java]: "log messages" --> "messages"

File: core/src/main/java/nu/yona/server/subscriptions/service/UserService.java [hunk 2]
```diff
@@ -278,14 +277,14 @@ public User addUserCreatedOnBuddyRequest(UserDTO buddyUser, String tempPassword)
 
 		// use a separate transaction to commit within the crypto session
 		UUID savedUserID = CryptoSession.execute(Optional.of(tempPassword), null, () -> transactionHelper
-				.executeInNewTransaction(() -> addUserCreatedOnBuddyRequestInSubTransaction(buddyUser).getID()));
+				.executeInNewTransaction(() -> addUserCreatedOnBuddyRequestInSubtransaction(buddyUser).getID()));
```
Comment (Bert-R) [inline core/src/main/java/nu/yona/server/subscriptions/service/UserService.java]: Eye for detail!

**Comment by aristotelos** [comment]
Corrected. The only thing I am still wondering about is how this concurrent call to `getReceivedMessageEntities` could ever happen. The only callers are the `MessageController` itself and the method `BuddyService.processPossiblePendingBuddyResponseMessage` (notice that those BuddyConnectResponseMessages are sent anonymous so that they do not need the transfer of direct messages). This is called by `BuddyService.processPossiblePendingBuddyResponseMessages`. The `foreach` there might be the cause but that is only the case if you get responses from two buddies at once which is rare. The only two callers to those methods are again `deleteUser` and `removeBuddy`.
So what I think is that they have been testing at Mobiquity with one account and multiple devices..

**Comment by Bert-R** [comment]
It's also quite possible that they send two concurrent requests, for instance one for all messages and one for the unread ones (to update the count).

**Pull request merged**